### PR TITLE
Lightning: optimize the estimated size

### PR DESCRIPTION
**Pull request by okJiang**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/41942

Problem Summary:

### What is changed and how it works?

The compression ratio for TiKV is typically around 3 to 1. 

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/lightning/restore/restore.go [hunk 1]
```diff
@@ -115,7 +115,8 @@ const (
 		status  VARCHAR(32) NOT NULL,
 		state   TINYINT(1) NOT NULL DEFAULT 0 COMMENT '0: normal, 1: exited before finish',
 		source_bytes BIGINT(20) UNSIGNED NOT NULL DEFAULT 0,
-		cluster_avail BIGINT(20) UNSIGNED NOT NULL DEFAULT 0,
```
Comment (dsdashun) [inline br/pkg/lightning/restore/restore.go:118]: For existing clutters，this table has already been created, and the table will only be created if the table doesn't exist. This means the table schema will remain the old one without those two new fields for those existing clusters.

File: br/pkg/lightning/restore/get_pre_info.go [hunk 2]
```diff
@@ -561,6 +561,10 @@ func (p *PreRestoreInfoGetterImpl) EstimateSourceDataSize(ctx context.Context, o
 		}
 	}
 
+	if !isTiDBBackend(p.cfg) {
+		sizeWithIndex = sizeWithIndex / 3
```
Comment (dsdashun) [inline br/pkg/lightning/restore/get_pre_info.go:565]: Can you add a comment here describing why size should be divided by 3 ?

File: br/pkg/lightning/restore/get_pre_info.go [hunk 3]
```diff
@@ -561,6 +561,10 @@ func (p *PreRestoreInfoGetterImpl) EstimateSourceDataSize(ctx context.Context, o
 		}
 	}
 
+	if !isTiDBBackend(p.cfg) {
```
Comment (gozssky) [inline br/pkg/lightning/restore/get_pre_info.go:564]:
```suggestion
	if isLocalBackend(p.cfg) {
```

File: br/pkg/lightning/restore/precheck_impl.go [hunk 4]
```diff
@@ -153,13 +162,23 @@ func (ci *clusterResourceCheckItem) Check(ctx context.Context) (*CheckResult, er
...
 		theResult.Passed = false
-		theResult.Message = fmt.Sprintf("Cluster doesn't have enough space, available is %s, but we need %s",
-			units.BytesSize(float64(clusterAvail)), units.BytesSize(float64(estimateSize)))
-	} else {
-		theResult.Message = fmt.Sprintf("Cluster available is rich, available is %s, we need %s",
-			units.BytesSize(float64(clusterAvail)), units.BytesSize(float64(estimateSize)))
+		theResult.Message += fmt.Sprintf("The estimated storage space required by the TiKV cluster is %s but the actual storage space is %s.",
+			units.BytesSize(float64(tikvAvail)), units.BytesSize(float64(estimateSize)))
+	}
+	if estimateSize > tiflashAvail {
```
Comment (gozssky) [inline br/pkg/lightning/restore/precheck_impl.go:175]: What happens if a table has no tiflash replica or a cluster has no tiflash nodes?

File: br/pkg/lightning/restore/precheck_impl.go [hunk 5]
```diff
@@ -153,13 +162,23 @@ func (ci *clusterResourceCheckItem) Check(ctx context.Context) (*CheckResult, er
 		return nil, errors.Trace(err)
 	}
 	estimateSize := clusterSource * replicaCount
```
Comment (dsdashun) [inline br/pkg/lightning/restore/precheck_impl.go:164]: Here, the estimated size is multiplied by the number of replicas.  However, you have divided the available size of the entire cluster into TiKV size and TiFlash size.  There may be false alarms.  For example, suppose there's only one TiFlash node in a cluster, and it is sufficient to store one replica.  However, the check indicates that this node is not large enough to store multiple replicas.  In this situation, the check should pass IMO.

File: br/pkg/lightning/restore/precheck_impl.go (hunk 4 above)
Comment (okJiang) [reply br/pkg/lightning/restore/precheck_impl.go:175]: thanks for your review, I will add these check

File: br/pkg/lightning/restore/restore.go (hunk 1 above)
Comment (okJiang) [reply br/pkg/lightning/restore/restore.go:118]: Lightning is once-only used. It is fine?

File: br/pkg/lightning/restore/precheck_impl.go (hunk 5 above)
Comment (okJiang) [reply br/pkg/lightning/restore/precheck_impl.go]: After https://github.com/pingcap/tidb/pull/41943/commits/263bba7236e98d724bedbc692c4c85508ac309e0 and https://github.com/pingcap/tidb/pull/41943/commits/a898ff616020d515ab4a48313c17058091b4994f, The `Tiflash` check should be same as `TiKV`, I think it be able to effect the result.

File: br/pkg/lightning/restore/precheck_impl.go (hunk 4 above)
Comment (okJiang) [reply br/pkg/lightning/restore/precheck_impl.go]: add it in https://github.com/pingcap/tidb/commit/263bba7236e98d724bedbc692c4c85508ac309e0 and https://github.com/pingcap/tidb/commit/a898ff616020d515ab4a48313c17058091b4994f. ptal~

File: br/pkg/lightning/restore/get_pre_info.go [hunk 6]
```diff
@@ -545,26 +556,35 @@ func (p *PreRestoreInfoGetterImpl) EstimateSourceDataSize(ctx context.Context, o
 					tbl.IndexRatio = sampledIndexRatio
 					tbl.IsRowOrdered = isRowOrderedFromSample
 
+					// if sample data failed due to max-error, fallback to use source size
```
Comment (gozssky) [inline br/pkg/lightning/restore/get_pre_info.go:559]: Can `tbl.IndexRatio` be zero?

File: br/pkg/lightning/restore/precheck_impl.go [hunk 7]
```diff
@@ -152,14 +166,26 @@ func (ci *clusterResourceCheckItem) Check(ctx context.Context) (*CheckResult, er
...
+
+	if estimateTikvSize > tikvAvail {
 		theResult.Passed = false
-		theResult.Message = fmt.Sprintf("Cluster doesn't have enough space, available is %s, but we need %s",
-			units.BytesSize(float64(clusterAvail)), units.BytesSize(float64(estimateSize)))
-	} else {
-		theResult.Message = fmt.Sprintf("Cluster available is rich, available is %s, we need %s",
-			units.BytesSize(float64(clusterAvail)), units.BytesSize(float64(estimateSize)))
+		theResult.Message += fmt.Sprintf("The estimated storage space required by the TiKV cluster is %s but the actual storage space is %s.",
+			units.BytesSize(float64(tikvAvail)), units.BytesSize(float64(estimateTikvSize)))
```
Comment (gozssky) [inline br/pkg/lightning/restore/precheck_impl.go:180]: Is this order wrong?

File: br/pkg/lightning/restore/precheck_impl.go [hunk 8]
```diff
@@ -152,14 +166,26 @@ func (ci *clusterResourceCheckItem) Check(ctx context.Context) (*CheckResult, er
...
+	}
+	if tiflashAvail > 0 && estimateTifalshSize > tiflashAvail {
+		theResult.Passed = false
+		theResult.Message += fmt.Sprintf(" The estimated storage space required for the Tiflash cluster is %s, but the actual storage space is %s.",
+			units.BytesSize(float64(tiflashAvail)), units.BytesSize(float64(estimateTifalshSize)))
```
Comment (gozssky) [inline br/pkg/lightning/restore/precheck_impl.go:185]: ditto

File: br/pkg/lightning/restore/get_pre_info.go (hunk 6 above)
Comment (okJiang) [reply br/pkg/lightning/restore/get_pre_info.go:559]: Seems it is a bug before....

Comment (okJiang) [reply br/pkg/lightning/restore/get_pre_info.go:559]: When `sampleDataFromTable` failed, `sampledIndexRatio/tbl.IndexRatio` will be zero.

File: br/pkg/lightning/restore/precheck_impl.go (hunk 7 above)
Comment (okJiang) [reply br/pkg/lightning/restore/precheck_impl.go]: yes, fixed in https://github.com/pingcap/tidb/pull/41943/commits/2da4928b04bf19365090cc061d70ad778aa4d0f9

File: br/pkg/lightning/restore/get_pre_info.go [hunk 9]
```diff
@@ -54,6 +54,9 @@ import (
 	"golang.org/x/exp/maps"
 )
 
+// compressionRatio is the tikv/tiflash's compression ratio
+const compressionRatio = float64(1) / 3
```
Comment (okJiang) [inline br/pkg/lightning/restore/get_pre_info.go:58]: After asked tiflash member, the size of tikv is similar with tiflash for one replica

File: br/pkg/lightning/importer/get_pre_info.go [hunk 10]
```diff
@@ -545,26 +556,35 @@ func (p *PreImportInfoGetterImpl) EstimateSourceDataSize(ctx context.Context, op
 					tbl.IndexRatio = sampledIndexRatio
 					tbl.IsRowOrdered = isRowOrderedFromSample
 
+					// if sample data failed due to max-error, fallback to use source size
 					if tbl.IndexRatio > 0 {
-						sizeWithIndex += int64(float64(tbl.TotalSize) * tbl.IndexRatio)
-					} else {
-						// if sample data failed due to max-error, fallback to use source size
-						sizeWithIndex += tbl.TotalSize
+						tableSize = int64(float64(tbl.TotalSize) * tbl.IndexRatio)
```
Comment (gozssky) [inline br/pkg/lightning/importer/get_pre_info.go:561]: Can we remove the `if tbl.IndexRatio > 0`?  It seems `tbl.IndexRatio` can not be zero.

File: br/pkg/lightning/importer/table_import_test.go [hunk 11]
```diff
@@ -99,6 +99,10 @@ func (s *tableRestoreSuiteBase) setupSuite(t *testing.T) {
 	core, err := ddl.MockTableInfo(se, node.(*ast.CreateTableStmt), 0xabcdef)
 	require.NoError(t, err)
 	core.State = model.StatePublic
+	core.TiFlashReplica = &model.TiFlashReplicaInfo{
```
Comment (gozssky) [inline br/pkg/lightning/importer/table_import_test.go:102]: Please add a test case for multiple tables with different tiflash replicas.

**Review by gozssky — COMMENTED** [review]
Rest LGTM.

File: br/pkg/lightning/importer/precheck_impl.go [hunk 12]
```diff
@@ -152,14 +166,26 @@ func (ci *clusterResourceCheckItem) Check(ctx context.Context) (*CheckResult, er
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
-	estimateSize := clusterSource * replicaCount
-	if estimateSize > clusterAvail {
+	estimateTikvSize := tikvSourceSize * replicaCount
+	// note: tiflashSourceSize contains replicaCount
```
[the remark this answers is not in the document]
Comment (okJiang) [reply br/pkg/lightning/importer/precheck_impl.go]: yes, we can https://github.com/pingcap/tidb/pull/41943/commits/931f006043591ae0ec7cb22e190013570b15cfad

File: br/pkg/lightning/importer/get_pre_info.go (hunk 10 above)
Comment (okJiang) [reply br/pkg/lightning/importer/get_pre_info.go]: It is removed in https://github.com/pingcap/tidb/pull/41943/commits/931f006043591ae0ec7cb22e190013570b15cfad

File: br/pkg/lightning/importer/table_import_test.go (hunk 11 above)
Comment (okJiang) [reply br/pkg/lightning/importer/table_import_test.go:99]: added in https://github.com/pingcap/tidb/pull/41943/commits/931f006043591ae0ec7cb22e190013570b15cfad

**Comment by okJiang** [comment]
ptal again~ @dsdashun @gozssky

**Review by gozssky — APPROVED** [review]

**Comment by gozssky** [comment]
@okJiang Did you do any tests on real clusters?

**Comment by okJiang** [comment]
> @okJiang Did you do any tests on real clusters?

Yes, but more detailed testing is underway.

> resp="{\"check_results\":[{\"check_item\":\"CHECK_ITEM_TARGET_CLUSTER_SIZE\", \"has_passed\":false, \"msg\":\"TiKV requires more storage space. Estimated required size: 312.6GiB. Actual size: 271.6GiB. Please increase storage to prevent import task failures.\"}

**Review by dsdashun — APPROVED** [review]

**Pull request merged**